File: dynamic_programming/longest_common_subsequence.rb
# Given two sequences, find the length of the longest subsequence present in both of them.
# A subsequence is a sequence that appears in the same relative order, but not necessarily contiguous.
# For example, “abc”, “abg”, “bdf”, “aeg”, ‘”acefg”, .. etc are subsequences of “abcdefg”.

# SOLUTION:
# Pseudocode
# I)
# We will first construct a 2D table from inputs string1 and string2 with a helper function common_subsequence_table;
# we will build this table in a bottom up way.
# The elements of the table, table[i][j], will contain the length of the longest
# common subsequence of string1[0..i-1] and strin2[0..j-1]
# The helper function, common_subsequence_table, returns this table
# The table will be built using the formula below:
#  table[i][j] = 0 if i = 0 or j = 0
#  table[i][j] = table[i-1][j-1] + 1 if i > 0, j > 0, and string1[i] = string2[j]
#  table[i][j] = max(table[i, j-1], table[i-1, j]) if i, j > 0 and tring1[i] != string2[j]

# II)
def longest_common_subsequence(string1, string2)
  table = common_subsequence_table(string1, string2)
  result = ""
  i = string1.length
  j = string2.length
  # we start from right bottom corner of table and 1 by 1 we store the common characters (of string1 and string2) into result.
  while i > 0 && j > 0
     # if current character in string1 and string2 are the same, we add it to result
    if string1[i-1] == string2[j-1]
      result.insert(0, string1[i-1]) # put current character at begining of result
      i -= 1 # decrement both i and j
      j -= 1
      # If the 2 characters are not the same, find the largest of 2 elements (left or up) from table and go into that direction
    elsif table[i-1][j] > table[i][j-1]
      i -= 1
    else
      j -= 1
    end
  end
  result
end

def common_subsequence_table(string1, string2)
  len1 = string1.length
  len2 = string2.length
  table = Array.new(len1+1) {Array.new(len2+1)}

  i = 0
  while i <= len1
    j = 0
    while j <= len2
      if i == 0 || j == 0
        table[i][j] = 0
      elsif string1[i-1] == string2[j-1]
        table[i][j] = table[i-1][j-1] + 1
      else
        table[i][j] = maximum(table[i][j-1], table[i-1][j])
      end
      j += 1
    end
    i += 1
  end
  table
end

def maximum(a, b)
  a > b ? a : b
end

p longest_common_subsequence("ABCBDAB", "BDCABC") == "BDAB"
p longest_common_subsequence("abc", "d") == ""
p longest_common_subsequence("ab", "a") == "a"
p longest_common_subsequence("ABCDGH", "AEDFHR") == "ADH"
p longest_common_subsequence("AGGTAB", "GXTXAYB") == "GTAB"

File: Problems/Java/LongestCommonSubsequence.java
/*
Longest Common Subsequence
From https://www.geeksforgeeks.org/longest-common-subsequence-dp-4/

Given two sequences, find the length of longest subsequence present in both of them.
A subsequence is a sequence that appears in the same relative order, but not necessarily contiguous.
For example, “abc”, “abg”, “bdf”, “aeg”, ‘”acefg”, .. etc are subsequences of “abcdefg”.
So a string of length n has 2^n different possible subsequences.

It is a classic computer science problem, the basis of diff (a file comparison program that outputs
the differences between two files), and has applications in bioinformatics.

Examples:
LCS for input Sequences “ABCDGH” and “AEDFHR” is “ADH” of length 3.
LCS for input Sequences “AGGTAB” and “GXTXAYB” is “GTAB” of length 4.
*/

public class LongestCommonSubsequence {

  // Recursive solution: Exponential runtime
  public static int longestCommonSubsequence(String s, String t) {
    if (s == null || t == null || s.length() == 0 || t.length() == 0) {
      return 0;
    }
    return traverse(s, t, s.length(), t.length());
  }

  public static int traverse(String s, String t, int m, int n) {
    if (m == 0 || n == 0) {
      return 0;
    } else if (s.charAt(m-1) == t.charAt(n-1)) {
      return 1 + traverse(s, t, m-1, n-1);
    } else {
      int trimLeft = traverse(s, t, m-1, n);
      int trimRight = traverse(s, t, m, n-1);
      return Math.max(trimLeft, trimRight);
    }
  }

  // Dynamic Programming solutioin:
  public static int longestCommonSubsequenceDP(String s, String t) {
    if (s == null || t == null || s.length() == 0 || t.length() == 0) {
      return 0;
    }

    int m = s.length(), n = t.length();
    int[][] dp = new int[m+1][n+1];

    for (int i = 0; i <= m; i++) {
      for (int j = 0; j <= n; j++) {
        if (i == 0 || j == 0) {
          dp[i][j] = 0;
        } else if (s.charAt(i-1) == t.charAt(j-1)) {
          dp[i][j] = 1 + dp[i-1][j-1];
        } else {
          dp[i][j] = Math.max(dp[i-1][j], dp[i][j-1]);
        }
      }
    }

    return dp[m][n];
  }

  public static void main(String[] args) {
    
    System.out.println("Using recursive solution:");
    System.out.println("longestCommonSubsequence(\"ABCDGH\", \"AEDFHR\") = " + longestCommonSubsequence("ABCDGH", "AEDFHR")); // 3 -> "ADH"
    System.out.println("longestCommonSubsequence(\"AGGTAB\", \"GTAB\") = " + longestCommonSubsequence("AGGTAB", "GTAB")); // 4 -> "ADH"

    System.out.println("\nUsing Dynamic Programming solution:");
    System.out.println("longestCommonSubsequenceDP(\"ABCDGH\", \"AEDFHR\") = " + longestCommonSubsequenceDP("ABCDGH", "AEDFHR"));
    System.out.println("longestCommonSubsequenceDP(\"AGGTAB\", \"GTAB\") = " + longestCommonSubsequenceDP("AGGTAB", "GTAB"));
  }
}
